backoffcache.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/libp2p/go-libp2p-peerstore/addr"
+)
+
+// BackoffDiscovery is an implementation of discovery that caches peer data and attenuates repeated queries
+type BackoffDiscovery struct {
+	disc         discovery.Discovery
+	stratFactory BackoffFactory
+	peerCache    map[string]*backoffCache
+	peerCacheMux sync.RWMutex
+}
+
+func NewBackoffDiscovery(disc discovery.Discovery, stratFactory BackoffFactory) (discovery.Discovery, error) {
+	return &BackoffDiscovery{
+		disc:         disc,
+		stratFactory: stratFactory,
+		peerCache:    make(map[string]*backoffCache),
+	}, nil
+}
+
+type backoffCache struct {
+	nextDiscover time.Time
+	prevPeers    map[peer.ID]peer.AddrInfo
+
+	peers      map[peer.ID]peer.AddrInfo
+	sendingChs map[chan peer.AddrInfo]int
+
+	ongoing bool
+	strat   BackoffStrategy
+	mux     sync.Mutex
+}
+
+func (d *BackoffDiscovery) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	return d.disc.Advertise(ctx, ns, opts...)
+}
+
+func (d *BackoffDiscovery) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get cached peers
+	d.peerCacheMux.RLock()
+	c, ok := d.peerCache[ns]
+	d.peerCacheMux.RUnlock()
+
+	/*
+		Overall plan:
+		If it's time to look for peers, look for peers, then return them
+		If it's not time then return cache
+		If it's time to look for peers, but we have already started looking. Get up to speed with ongoing request
+	*/
+
+	// Setup cache if we don't have one yet
+	if !ok {
+		pc := &backoffCache{
+			nextDiscover: time.Time{},
+			prevPeers:    make(map[peer.ID]peer.AddrInfo),
+			peers:        make(map[peer.ID]peer.AddrInfo),
+			sendingChs:   make(map[chan peer.AddrInfo]int),
+			strat:        d.stratFactory(),
+		}
+		d.peerCacheMux.Lock()
+		c, ok = d.peerCache[ns]
+
+		if !ok {
+			d.peerCache[ns] = pc
+			c = pc
+		}
+
+		d.peerCacheMux.Unlock()
+	}
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	findPeers := !ok
+	timeExpired := false
+	if !findPeers {
+		timeExpired = time.Now().After(c.nextDiscover)
+		findPeers = timeExpired && !c.ongoing
+	}
+
+	// If we should find peers then setup a dispatcher channel for dispatching incoming peers
+	if findPeers {
+		pch, err := d.disc.FindPeers(ctx, ns, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		c.ongoing = true
+		go findPeerDispatcher(ctx, c, pch)
+		// If it's not yet time to search again then return cached peers
+	} else if !timeExpired {
+		chLen := options.Limit
+
+		if chLen == 0 {
+			chLen = len(c.prevPeers)
+		} else if chLen > len(c.prevPeers) {
+			chLen = len(c.prevPeers)
+		}
+		pch := make(chan peer.AddrInfo, chLen)
+		for _, ai := range c.prevPeers {
+			pch <- ai
+		}
+		close(pch)
+		return pch, nil
+	}
+
+	// Setup receiver channel for receiving peers from ongoing requests
+	evtCh := make(chan peer.AddrInfo, 32)
```

Comment:
maybe extract 32 out into an exported `var`

Reply:
any thoughts on using a global var vs a custom discovery option `options.Other["backoff : evtChSize"]=32`?

Reply:
yeah, that's fine. that key is pretty odd, i would definitely export is as a public var so others might reference it more easily

Reply:
did you just want a variable for the internal peer channel buffer size or for the one we return as well?

They're currently both 32, but they don't inherently have to be the same.

Reply:
used options instead of global variables (figured there's no reason why this should have to be global when it's possible that the caches won't be singletons). If this turns out to be something people actually configure we can add per namespace/request configurations (e.g. using the options as described above)